File: Visual Studio Projects/OverprintSimulation/CustomSpotMergingCpp/CustomSpotMerging.cpp

/* -----------------------------------------------------------------------
 * <copyright file="CustomSpotMerge.cpp" company="Hybrid Software Ltd">
 *  Copyright (c) 2025 Hybrid Software Ltd. All rights reserved.
 * </copyright>
 * <summary>
 *  This example is provided on an "as is" basis and without warranty of any kind.
 *  Hybrid Software Ltd. does not warrant or make any representations
 *  regarding the use or results of use of this example.
 * </summary>
 * -----------------------------------------------------------------------
 */

#include <filesystem>
#include <iostream>

#include <jawsmako/jawsmako.h>
#include <jawsmako/pdfoutput.h>
#include <jawsmako/pdfinput.h>

using namespace JawsMako;
using namespace EDL;

const U8String TEST_FILES_PATH = R"(..\..\..\..\TestFiles\)";

int main()
{
    try
    {
        const auto mako = IJawsMako::create();
        IJawsMako::enableAllFeatures(mako);

        // Load the document
        auto documentAssembly = IPDFInput::create(mako)->open(TEST_FILES_PATH + "Robots Plus Process Colors.pdf");
        auto document = documentAssembly->getDocument();
        for (uint32 i = 0; i < document->getNumPages(); i++)
        {
            // Set image dimensions + colorspace
            const auto fixedPage = document->getPage(i)->getContent();
            const auto bounds = FRect(0, 0, fixedPage->getWidth(), fixedPage->getHeight());
            constexpr double resolution = 150;
            const auto pixelWidth = static_cast<uint32_t>(lround(bounds.dX / 96.0 * resolution));
            const auto pixelHeight = static_cast<uint32_t>(lround(bounds.dY / 96.0 * resolution));
            const auto cmyk = IDOMColorSpaceDeviceCMYK::create(mako); // colorspace must be cmyk for spot merging (we can convert back to rgb later if necessary)

            // Get spot lists
            const auto spots = IRendererTransform::inkInfoToColorantInfo(mako, IRendererTransform::findInks(mako, fixedPage), cmyk);
            CSpotColorNames spotNames;
            for (auto& spot : spots) 
                spotNames.append(spot.name);

            // Prepare frame buffers
            const auto numProcess = cmyk->getNumComponents();
            const auto numSpots = spots.size();
            const auto numBuffers = numProcess + numSpots;
            std::vector<std::vector<uint8_t>> buffers(numBuffers);
            auto frameBuffers = IJawsRenderer::CFrameBufferInfoVect(numBuffers);
            for (unsigned bufferIndex = 0; bufferIndex < numBuffers; ++bufferIndex)
            {
                buffers[bufferIndex].resize(pixelHeight * pixelWidth);
                frameBuffers[bufferIndex].buffer = buffers[bufferIndex].data();
                frameBuffers[bufferIndex].rowStride = static_cast<int>(pixelWidth);
                frameBuffers[bufferIndex].pixelStride = 0;
            }

            // Render using renderSeparationsToFrameBuffers()
            auto renderer = IJawsRenderer::create(mako);
            renderer->renderSeparationsToFrameBuffers(
                fixedPage,
                8,
                true,
                pixelWidth,
                pixelHeight,
                cmyk,
                frameBuffers,
                0,
                bounds,
                spotNames,
                IOptionalContentPtr(),
                eOCEPrint // Non printable layers will be off
            );


            // write the buffers to an image
            IImageFrameWriterPtr frameWriter;
            auto image = IDOMRawImage::createWriterAndImage(mako, frameWriter, cmyk,
                pixelWidth, pixelHeight,
                8, resolution, resolution);

            // Get spot components 
            CEDLVector<CFloatVect> components(numSpots);
            for (uint32_t s = 0; s < numSpots; s++)
                components[s] = CFloatVect({
                    spots[s].components[0],
                    spots[s].components[1],
                    spots[s].components[2],
                    spots[s].components[3]
                    });

            // Merge spots into process
            CEDLVector<uint8_t*> outPtrs(numProcess);
            CEDLVector<uint8_t*> inPtrs(numSpots);
            constexpr float inv255 = 1.0f / 255.0f;

            for (uint32_t y = 0; y < pixelHeight; ++y)
            {
                for (uint32_t j = 0; j < numProcess; ++j)
                    outPtrs[j] = static_cast<uint8_t*>(frameBuffers[j].buffer) + y * frameBuffers[j].rowStride;

                for (uint32_t k = 0; k < numSpots; ++k)
                    inPtrs[k] = static_cast<uint8_t*>(frameBuffers[numProcess + k].buffer) + y * frameBuffers[numProcess + k].rowStride;

                auto scanline = std::vector<uint8_t>(pixelWidth * numProcess);

                for (uint32_t x = 0; x < pixelWidth; ++x)
                    for (uint32_t c = 0; c < 4; ++c)
                    {
                        scanline[x * numProcess + c] = outPtrs[c][x];
                        for (uint32_t s = 0; s < numSpots; ++s)
                        {
                            float spotVal = inPtrs[s][x] * inv255;
                            float currentVal = scanline[x * numProcess + c] * inv255;
                            float newVal = 1.0f - ((1.0f - components[s][c] * spotVal) * (1.0f - currentVal));
                            scanline[x * numProcess + c] = static_cast<uint8_t>(newVal * 255.0f + 0.5f);
                        }
                    }
                frameWriter->writeScanLine(scanline.data());
            }

            frameWriter->flushData();

            //now we can convert back to RGB if needed
            const auto filteredImage = IDOMFilteredImage::create(mako, image, IDOMImageColorConverterFilter::create(mako, IDOMColorSpaceDeviceRGB::create(mako), eRelativeColorimetric, eBPCDefault));

            // save the output to jpeg file
            auto outputJPEG = "output_" + std::to_string(i) + ".jpg";
            IDOMJPEGImage::encode(mako, static_cast<IDOMImagePtr>(filteredImage), IOutputStream::createToFile(mako, outputJPEG.c_str()));
        }
    }
    catch (IError& e)
    {
        const String errorFormatString = getEDLErrorString(e.getErrorCode());
        std::wcerr << L"Exception thrown: " << e.getErrorDescription(errorFormatString) << '\n';
        return static_cast<int>(e.getErrorCode());
    }
    catch (std::exception& e)
    {
        std::wcerr << L"std::exception thrown: " << e.what() << '\n';
        return 1;
    }

    return 0;
}

File: Visual Studio Projects/OverprintSimulation/CustomSpotMergingCS/CustomSpotMerging.cs
// -----------------------------------------------------------------------
// <copyright file="CustomSpotMerge.cs" company="Hybrid Software">
//   Copyright (c) 2025 Hybrid Software.
// </copyright>
// <summary>
//   Example provided "as is", without warranty of any kind.
// </summary>
// -----------------------------------------------------------------------

using JawsMako;

class CustomSpotMergingCS
{
    private const string TestFilesPath = @"..\..\..\..\..\..\TestFiles\";

    static int Main()
    {
        try
        {
            using var mako = IJawsMako.create();
            IJawsMako.enableAllFeatures(mako);

            // Load the document
            using var assembly = IPDFInput.create(mako).open(Path.Combine(TestFilesPath, "Robots Plus Process Colors.pdf"));
            using var document = assembly.getDocument();

            for (uint pageIndex = 0; pageIndex < document.getNumPages(); pageIndex++)
            {
                using var fixedPage = document.getPage(pageIndex).getContent();

                // Page bounds & raster geometry
                using var bounds = new FRect(0, 0, fixedPage.getWidth(), fixedPage.getHeight());
                const double resolution = 150.0;
                var pixelWidth = (uint)Math.Round(bounds.dX / 96.0 * resolution);
                var pixelHeight = (uint)Math.Round(bounds.dY / 96.0 * resolution);

                // Colorspace must be CMYK for spot merging (we can convert to RGB later)
                using var cmyk = IDOMColorSpaceDeviceCMYK.create(mako);

                // ----- Find inks and build spot list (names + CMYK components) -----
                using var spots = IRendererTransform.inkInfoToColorantInfo(mako, IRendererTransform.findInks(mako, fixedPage), cmyk);
                using var spotNames = new CEDLVectWString();

                foreach (var spot in spots.toVector())
                    spotNames.append(spot.name);

                int numProcess = cmyk.getNumComponents(); // 4 (CMYK)
                var numSpots = (int)spots.size();
                var numBuffers = numProcess + numSpots;

                // ----- Prepare frame buffers for renderSeparationsToFrameBuffers() -----
                var buffers = new byte[numBuffers][];
                using var fb = new CEDLVectCFrameBufferInfo();

                for (var i = 0; i < numBuffers; i++)
                {
                    buffers[i] = new byte[pixelHeight *  pixelWidth];

                    var info = new IJawsRenderer.CFrameBufferInfo
                    {
                        bufferOfs = 0,
                        rowStride = (int)pixelWidth,   // pixels per row (1 byte per pixel)
                        pixelStride = 0                // tightly packed
                    };
                    fb.append(info);
                }

                // ----- Render true separations (process first, then spots) -----
                using var renderer = IJawsRenderer.create(mako);
                renderer.renderSeparationsToFrameBuffers(
                    fixedPage,
                    8,                    // bits per component per plate
                    true,                 // antialiased
                    pixelWidth,
                    pixelHeight,
                    cmyk,                 // process space -> first 4 plates (C,M,Y,K)
                    buffers,
                    fb,
                    0,                    // bandMemorySize (0 = no banding)
                    bounds,
                    spotNames,            // which spots to keep (all found)
                    IOptionalContent.Null(),
                    eOptionalContentEvent.eOCEPrint // Non-printable layers will be off
                );

                // ----- Create writer and image -----
                using var pair = IDOMRawImage.createWriterAndImage(
                    mako,
                    cmyk,
                    pixelWidth,
                    pixelHeight,
                    8,
                    resolution, resolution
                );

                using var frameWriter = pair.frameWriter;

                // ----- Get spot components of spots to merge -----
                using var components = new CEDLVectVectFloat();
                for (uint i = 0; i < numSpots; i++)
                    components.append(new CEDLVectFloat(new StdVectFloat()
                    {
                        spots[i].components[0],
                        spots[i].components[1],
                        spots[i].components[2],
                        spots[i].components[3]
                    }));

                // ----- Merge each spot component with the process buffer values and write the result to the scanline -----
                const float inv255 = 1.0f / 255.0f;
                for (uint y = 0; y < pixelHeight; y++)
                {
                    var rowStart = (int)(y * pixelWidth);
                    var scanline = new byte[pixelWidth * numProcess];

                    for (uint x = 0; x < pixelWidth; ++x)
                    {
                        for (uint c = 0; c < numProcess; ++c)
                        {
                            scanline[x * numProcess + c] = buffers[c][rowStart + x];
                            for (uint i = 0; i < numSpots; ++i)
                            {
                                var spotVal = buffers[numProcess + i][rowStart + x] * inv255;
                                var currentVal = scanline[x * numProcess + c] * inv255;
                                var newVal = 1.0f - (1.0f - components[i][c] * spotVal) * (1.0f - currentVal);
                                scanline[x * numProcess + c] = (byte)(newVal * 255.0f + 0.5f);
                            }
                        }
                    }
                    frameWriter.writeScanLine(scanline);
                }

                frameWriter.flushData();

                // ----- Convert to RGB (optional) and save as a JPEG (or other image type) -----
                using var rgb = IDOMColorSpaceDeviceRGB.create(mako);
                using var cc = IDOMImageColorConverterFilter.create(mako, rgb, eRenderingIntent.eRelativeColorimetric, eBlackPointCompensation.eBPCDefault);
                using var filtered = IDOMFilteredImage.create(mako, pair.domImage, cc);

                var outJpeg = $"output_{pageIndex}.jpg";
                IDOMJPEGImage.encode(mako, filtered, IOutputStream.createToFile(mako, outJpeg));
                Console.WriteLine($"Wrote: {outJpeg}");
            }
        }
        catch (MakoException me)
        {
            Console.Error.WriteLine($"Mako error {me.m_errorCode}: {me.m_msg}");
            return 1;
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine(ex);
            return 1;
        }

        return 0;
    }
}